File: backend/app/schemas/watchlist.py
from __future__ import annotations

from pydantic import BaseModel


class WatchlistCompany(BaseModel):
    """A company the user wants to track."""
    name: str                   # Display name: "Netflix"
    slug: str = ""              # ATS board token: "netflix" (workday: "tenant/site")
    ats: str = ""               # "greenhouse" | "lever" | "ashby" | "workday" | "unknown"
    job_count: int = 0          # Last known open job count
    careers_url: str = ""       # Direct link to career page


class WatchlistAddRequest(BaseModel):
    """Request to add a company: a name, a careers link, or both."""
    name: str = ""
    url: str = ""


class WatchlistResponse(BaseModel):
    """Full watchlist for a profile.

    ``message`` is set when an add finished in an incomplete state (no job
    board found for the name), so the caller can ask for the careers link
    instead of presenting the inert entry as healthy.
    """
    profile: str
    companies: list[WatchlistCompany] = []
    message: str = ""

File: frontend/src/api/watchlist.ts
import { apiGet, apiPost, apiDelete } from '@/lib/api-client';

export interface WatchlistCompany {
  name: string;
  slug: string;
  ats: string;
  job_count: number;
  careers_url: string;
}

/* An add carries a name, a careers link, or both. A link resolves to its
   exact board token on the backend; name plus link completes an existing
   unfinished entry in place. */
export interface WatchlistAddPayload {
  name?: string;
  url?: string;
}

export interface WatchlistResponse {
  profile: string;
  companies: WatchlistCompany[];
  /* Set when an add finished without a confirmed board, so the UI can ask
     for the careers link instead of showing the entry as healthy. */
  message?: string;
}

export function getWatchlist(profile: string): Promise<WatchlistResponse> {
  return apiGet<WatchlistResponse>(`/profiles/${profile}/watchlist`);
}

export function addWatchlistCompany(
  profile: string,
  payload: WatchlistAddPayload,
): Promise<WatchlistResponse> {
  return apiPost<WatchlistResponse>(`/profiles/${profile}/watchlist`, {
    name: payload.name ?? '',
    url: payload.url ?? '',
  });
}

export function removeWatchlistCompany(profile: string, name: string): Promise<WatchlistResponse> {
  return apiDelete<WatchlistResponse>(`/profiles/${profile}/watchlist/${encodeURIComponent(name)}`);
}
